decorators.ts

```diff
@@ -4,36 +4,53 @@
 import Promise = require("bluebird");
 import fiberBootstrap = require("./fiber-bootstrap");
 
-export function exported(moduleName: string) {
-	return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => {
+export function exported(moduleName: string): any {
+	return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>): TypedPropertyDescriptor<any> => {
 		$injector.publicApi.__modules__[moduleName] = $injector.publicApi.__modules__[moduleName] || {};
-		$injector.publicApi.__modules__[moduleName][propertyKey] = (...args: any[]): any => {
-			return new Promise(function(onFulfilled : Function, onRejected: Function) {
-					let originalModule = $injector.resolve(moduleName);
-					let originalMethod: Function = originalModule[propertyKey];
-					let result: any;
-					try {
-						result = originalMethod.apply(originalModule, args);
-					} catch(err) {
-						onRejected(err);
-						return;
-					}
-
-					if(result && typeof result.wait === "function") {
-						fiberBootstrap.run(function () {
-							try {
-								let realResult = result.wait();
-								onFulfilled(realResult);
-							} catch(err) {
-								onRejected(err);
-							}
-						});
-					} else {
-						onFulfilled(result);
-					}
+		$injector.publicApi.__modules__[moduleName][propertyKey] = (...args: any[]): Promise<any>[] | Promise<any>=> {
+			let originalModule = $injector.resolve(moduleName);
+			let originalMethod: Function = originalModule[propertyKey];
+			let result: any;
+			try {
+				result = originalMethod.apply(originalModule, args);
+			} catch(err) {
+				let promise = new Promise(function(onFulfilled : Function, onRejected: Function) {
+					onRejected(err);
 				});
-		};
 
+				// TODO: Check if we should return [promise] or promise.
```

Comment:
We could... In case we know what to do - I'm not sure what should be returned in this case as we do not know to return type of the function.